kernel/src/sched/sched_class/fair.rs

```diff
@@ -0,0 +1,242 @@
+    }
+}
+
+impl Eq for Wrapper {}
+
+impl PartialOrd for Wrapper {
+    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+        Some(self.key().cmp(&other.key()))
+    }
+}
+
+impl Ord for Wrapper {
+    fn cmp(&self, other: &Self) -> cmp::Ordering {
+        self.key().cmp(&other.key())
+    }
+}
+
+struct Fair {
+    cpu: u32,
+    threads: BTreeMap<Wrapper, ()>,
+    load: u64,
+    num: u64,
+}
+
+impl core::fmt::Debug for Fair {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        writeln!(f, "Fair: load = {}, num = {}", self.load, self.num)?;
+        writeln!(f, "  threads: ")?;
+        for thread in self.threads.keys() {
+            let vr = thread.key();
+            writeln!(f, "    {vr:?}")?;
+        }
+        Ok(())
+    }
+}
+
+impl Fair {
+    fn new(cpu: u32) -> Self {
+        Self {
+            cpu,
+            threads: BTreeMap::new(),
+            load: 0,
+            num: 0,
+        }
+    }
+
+    fn period(&self) -> u64 {
+        let base_slice_clks = base_slice_clocks();
+        let min_period_clks = min_period_clocks();
+
+        let min_gran_clks = base_slice_clks * u64::from((1 + num_cpus()).ilog2());
+        (min_gran_clks * self.num).max(min_period_clks)
+    }
+
+    fn pop(&mut self, target_cpu: u32) -> Option<Arc<Thread>> {
+        let mut front = self.threads.lower_bound_mut(Bound::Unbounded);
+        let Wrapper(thread) = loop {
+            let (thread, _) = front.peek_next()?;
+            if !thread
+                .0
+                .task()
+                .upgrade()
+                .unwrap()
+                .schedule_info()
+                .cpu_affinity
+                .contains(target_cpu)
+            {
+                front.next().unwrap();
+                continue;
+            }
+
+            let (thread, _) = front.remove_next().unwrap();
+            break thread;
+        };
+
+        with_entity_mut(&thread, |vr: &mut VRuntime| vr.start = sched_clock());
+        Some(thread)
+    }
+}
+
+impl SchedClassRq for Fair {
+    fn kind(&self) -> SchedKind {
+        SchedKind::Fair
+    }
+
+    fn enqueue(&mut self, thread: Arc<Thread>) {
+        let vruntime: VRuntime = entity(&thread);
+        self.load += vruntime.weight;
+        self.num += 1;
+        self.threads.insert(Wrapper(thread), ());
+    }
+
+    fn dequeue(&mut self, thread: &Arc<Thread>) {
+        let vruntime: VRuntime = entity(thread);
+        self.load -= vruntime.weight;
+        self.num -= 1;
+    }
+
+    fn pick_next(&mut self) -> Option<Arc<Thread>> {
+        self.pop(self.cpu)
+    }
+
+    fn update_current(&mut self, cur: &Thread, flags: UpdateFlags) -> bool {
+        with_entity_mut(cur, |v: &mut VRuntime| match flags {
+            UpdateFlags::Yield => {
+                v.period_start = sched_clock();
+                true
+            }
+            UpdateFlags::Tick | UpdateFlags::Wait => v.tick(self.load, self.period()),
```

Comment:
The current thread is popped out before task switching as stated in `pick_next`, so the update here doesn't affect the integrity of `BTreeSet`. A debug assertion could be added to ensure the invariant, but it would dramatically affect the debug performance and slow down the CI.

Reply:
Ok. I will leave this issue open for now and reason about the correctness of the scheduling logic after necessary Rust docs are added and the code is polished for readability.